pex/pex_bootstrapper.py

```diff
@@ -56,28 +61,105 @@ def find_in_path(target_interpreter):
+    else:
+      # get all qualifying interpreters found in $PATH
+      interpreters = PythonInterpreter.all()
 
-  from .common import die
-  from .tracer import TRACER
+  return list(matched_interpreters(
+    interpreters, compatibility_constraints, meet_all_constraints=True))
 
-  target_python = ENV.PEX_PYTHON
+
+def _select_pex_python_interpreter(target_python, compatibility_constraints):
   target = find_in_path(target_python)
+
   if not target:
     die('Failed to find interpreter specified by PEX_PYTHON: %s' % target)
+  elif compatibility_constraints:
+    pi = PythonInterpreter.from_binary(target)
+    if not list(matched_interpreters([pi], compatibility_constraints, meet_all_constraints=True)):
+      die('Interpreter specified by PEX_PYTHON (%s) is not compatible with specified '
+          'interpreter constraints: %s' % (target, str(compatibility_constraints)))
+  if not os.path.exists(target):
+    die('Target interpreter specified by PEX_PYTHON %s does not exist. Exiting.' % target)
+  return target
+
+
+def _select_interpreter(pex_python_path, compatibility_constraints):
+  compatible_interpreters = find_compatible_interpreters(
+    pex_python_path, compatibility_constraints)
+
+  if not compatible_interpreters:
+    die('Failed to find compatible interpreter for constraints: %s'
+        % str(compatibility_constraints))
+  # TODO: https://github.com/pantsbuild/pex/issues/430
+  target = min(compatible_interpreters).binary
+
   if os.path.exists(target) and os.path.realpath(target) != os.path.realpath(sys.executable):
-    TRACER.log('Detected PEX_PYTHON, re-exec to %s' % target)
+    return target
+
+
+def maybe_reexec_pex(compatibility_constraints):
+  """
+  Handle environment overrides for the Python interpreter to use when executing this pex.
+
+  This function supports interpreter filtering based on interpreter constraints stored in PEX-INFO
+  metadata. If PEX_PYTHON is set in a pexrc, it attempts to obtain the binary location of the
+  interpreter specified by PEX_PYTHON. If PEX_PYTHON_PATH is set, it attempts to search the path for
+  a matching interpreter in accordance with the interpreter constraints. If both variables are
+  present in a pexrc, this function gives precedence to PEX_PYTHON_PATH and errors out if no
+  compatible interpreters can be found on said path. If neither variable is set, fall through to
+  plain pex execution using the currently executing Python interpreter.
+
+  :param compatibility_constraints: list of requirements-style strings that constrain the
+  Python interpreter to re-exec this pex with.
+
+  """
+  if ENV.SHOULD_EXIT_BOOTSTRAP_REEXEC:
+    return
+
+  selected_interpreter = None
+  with TRACER.timed('Selecting runtime interpreter based on pexrc', V=3):
+    if ENV.PEX_PYTHON and not ENV.PEX_PYTHON_PATH:
+      # preserve PEX_PYTHON re-exec for backwards compatibility
+      selected_interpreter = _select_pex_python_interpreter(ENV.PEX_PYTHON,
```

Comment:
I agree about the simplification. We discussed leaving it in as a way to give us flexibility when migrating the /etc/pexrc configs in prod and in MDE, as well as provide backwards interoperability for pex users who want to continue using a single python interpreter. However, IIRC, I think the main reason was to give us flexibility with the whole migration of /etc/pexrc and to give us a safety net in case we find that PEX_PYTHON_PATH ends up disastrously breaking things or causing other unexpected problems. 

I like the idea of a safety net and backwards compatibility (at least for now), and I think it wouldn't be a bad idea to keep it in for now until we have success using PPP internally, maybe with a reasonable subset of Python users successfully running and deploying py3 PEX both locally and in prod? What do you think? @UnrememberMe any thoughts here?